### PR TITLE
FSA22V2-238: Set up Airtable API connection

### PR DESCRIPTION
**Pull request by marissahuysentruyt**
### Description:

Add support for Airtable API. This database holds all of the data the site will need to showcase elements and code for the accessible components. 

The `.env.example` file shows what info would be needed in your `.env.local` file. The `airtable.js` file initializes the base/table, and contains the fetch function to call the API. Then, within `[id].page.jsx`, `getStaticProps` was updated to use the `id`s from the `param` objects in `getStaticPaths`, make the API call (`getComponentDetails`) with that `id`, and use the returned `Component Name` as the argument in the `Details` function to properly display the retrieved component's individual page heading and title. 

See Story: [FSA22V2-238](https://sparkbox.atlassian.net/browse/FSA22V2-238)

### Validation:

* [x] This PR has code changes, and our linters still pass.
* [x] This PR has new code, so new tests were added or updated, and they pass.

#### To Validate:

1. Make sure all PR Checks have passed (GitHub Actions, CircleCI, Code Climate, etc).
2. Pull down all related branches.
3. Run `npm i` to make sure you have `Airtable` and `dotenv `packages.
4. Create a `.env.local` file and include the following: 
```
AIRTABLE_API_KEY=<generate an API key from your Airtable account overview>
AIRTABLE_BASE_ID=<find the ID of your base in your Airtable API documentation>
AIRTABLE_TABLE_NAME=<your base's name>
```
5. In terminal: `npm run dev`.
6. In [localhost:3000](http://localhost:3000/), click on component card links to see individual pages.
7. Confirm heading matches the component link you followed, and is capitalized. 
8. Run `npm run lint` to confirm no linting errors. 
9. Run `npm run test` to confirm all tests pass.

---

### CODE REVIEW

File: src/pages/[id].page.jsx [hunk 1]
```diff
@@ -41,17 +43,40 @@ export default function Details() {
...
-});
+export async function getStaticProps({ params }) {
+  try {
+    const { id } = params;
+    // run the airtable function to get the data and save to the details variable.
+    const details = await getComponentDetails(id);
+    // pass that details variable and its data into props to use elsewhere
+    return {
+      props: {
+        componentName: details['Component Name'],
```
Comment (bsmrdel101) [inline src/pages/[id].page.jsx:75]: Nice Job! This was a really neat solution to getting Airtable data into the component.

File: src/styles/components/_footer.scss [hunk 2]
```diff
@@ -1,28 +1,27 @@
...
+    min-height: 20.75rem; /* 332px */
     display: flex;
     flex-direction: column;
     justify-content: space-between;
     align-items: center;
     gap: 0.5rem;
     text-align: center;
     padding: 3.75rem; /* 60px/16px */
-    font-size: 1.25rem; /* 20px/16px */
+    font-size: map.get($font-sizes, "sm"); /* 20px/16px */
```
Comment (bsmrdel101) [inline src/styles/components/_footer.scss:17]: Good job using the fonts and colors map.

File: src/styles/globals.scss [hunk 3]
```diff
@@ -1,67 +1,27 @@
...
-html,
-body {
-  padding: 0;
-  margin: 0;
-  background-color: map.get($colors, "white");
-  font-size: map.get($font-sizes, "sm");
-  font-family:
-    Oxygen,
-    -apple-system,
-    BlinkMacSystemFont,
```
Comment (bsmrdel101) [inline src/styles/globals.scss:23]: Cleaning up the _gloabls.scss was a good change 🏆

**Review by bsmrdel101 — APPROVED** [review]
Everything worked for me, and I was able to see the headings being rendered! However you should include the Jira card inside the PR description, and provide steps for getting the Airtable api data for the .env.local file in the validation instructions.

File: .env.example [hunk 4]
```diff
@@ -0,0 +1,3 @@
+AIRTABLE_API_KEY=
+AIRTABLE_BASE_ID=
+AIRTABLE_TABLE_NAME=
```
Comment (changangus) [inline .env.example:3]: Nit pick here, but if you decide to keep the table name hard coded in the `getComponentDetails` function you may want to remove this variable from the example file.

File: utils/airtable.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          }
+        },
+        (error) => {
+          if (error) {
+            reject(error);
+          }
+        },
+      );
+  });
+}
```
Comment (changangus) [inline utils/airtable.js:44]: Really appreciate the code comments! Nice work!

**Review by changangus — APPROVED** [review]
Hey @marissahuysentruyt, code looked good, everything was working great when I ran it and I was able to complete all the validation steps! The only thing is that there were a good amount of changes that were unrelated to the PR, just a reminder to avoid a lot of context changing in a single PR. Way to be proactive, but I would suggest creating new cards for unrelated issues you find while working on a PR. You're doing great work though, keep it up!

File: package.json [hunk 6]
```diff
@@ -27,6 +28,7 @@
     "@testing-library/jest-dom": "^5.16.5",
     "@testing-library/react": "^13.4.0",
     "axe-core": "^4.4.3",
+    "dotenv": "^16.0.3",
```
Comment (dustin-jw) [inline package.json:31]: I think we ended up not needing `dotenv`. We can go ahead an remove it. `npm uninstall --save-dev dotenv`

File: src/pages/[id].page.jsx [hunk 7]
```diff
@@ -41,17 +39,40 @@ export default function Details() {
...
+      },
+    };
+  } catch {
+    // throw new Error('No records matched the component name.');
```
Comment (dustin-jw) [inline src/pages/[id].page.jsx:75]: I call these types of comments "zombie code" since they can come back from the dead and wreak havoc. Let's get rid of this 🧟

File: src/pages/[id].page.test.jsx [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import '@testing-library/jest-dom';
+// import { render, screen } from '@testing-library/react';
+// import { act } from 'react-dom/test-utils';
```
Comment (dustin-jw) [inline src/pages/[id].page.test.jsx:3]: More zombie code 🧟 
In fact, this whole file isn't really doing anything super useful, so I'd be fine with removing it entirely. The individual components that make up the `[id]` page should be tested well enough to give us confidence that the whole page will work (and it's easy enough to check with our limited number of pages).

File: src/pages/[id].page.jsx [hunk 9]
```diff
@@ -1,14 +1,12 @@
...
-export default function Details() {
-  const router = useRouter();
-  const { id } = router.query;
+export default function Details({ componentName }) {
   return (
-    <Layout pageTitle={id}>
+    <Layout pageTitle={componentName}>
       <div>
-        <h1>{ id }</h1>
+        <h1 data-testid="heading-1">{componentName}</h1>
```
Comment (dustin-jw) [inline src/pages/[id].page.jsx:9]: Are we using this `testid` anywhere? I think we can remove it, and any tests looking for this element can look for specific text or an `h1` tag anyway.

File: src/styles/components/_component-card.scss [hunk 10]
```diff
@@ -28,4 +26,10 @@
     display: block;
     border-top: 0.25rem solid black;
   }
+
+  :focus {
+    outline: none;
+    border: 2px solid map.get($colors, "accent-teal");
+  }
 }
+
```
Comment (dustin-jw) [inline src/styles/components/_component-card.scss]: I think all of these `styles/*` changes are unrelated to this PR. Was this one of those other fixes that snuck in here? I would expect that no files from the `styles` folder would end up in this PR.

File: utils/airtable.test.jsx [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+];
+
+describe('findMatchingRecord', () => {
+  it('returns all of the data associated with a particular component', () => {
+    const mockComponentName = 'disclosure';
+    const matchingRecord = findMatchingRecord(
+      fakeAirtableData,
+      mockComponentName,
+    );
+    // eslint-disable-next-line no-underscore-dangle
```
Comment (dustin-jw) [inline utils/airtable.test.jsx:24]: For posterity, this `eslint-disable` comment is fine because Airtable is the one imposing their dangling underscores on us.

File: utils/airtable.test.jsx [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+describe('an unsuccessful API call', () => {
+  it('returns null when there is no match', async () => {
+    const mockComponentName = 'accordion';
+    const matchingRecord = findMatchingRecord(
+      fakeAirtableData,
+      mockComponentName,
+    );
+    expect(matchingRecord).toBeNull();
+  });
+});
```
Comment (dustin-jw) [inline utils/airtable.test.jsx:38]: Normally I would put this `it` statement in the first `describe` block, since it is also testing the same function. For organization, it's often handy to have multiple tests grouped together like that.

```js
describe('someFunction', () => {
  it('handles the happy path', () => {});

  it('handles the sad path'. () => {});
});
```

That said, these are great tests!

File: utils/findMatchingRecord.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+  const matchingRecord = records.find((record) => {
+    /* eslint-disable no-underscore-dangle */
+    if (record._rawJson.fields.Slug.toLowerCase() === componentName) {
+      return true;
+    }
+    return false;
+  });
+  // keep the `?? null` to ensure it returns null, instead of undefined
+  // if it can't find a matching record
+  return matchingRecord?._rawJson.fields ?? null;
```
Comment (dustin-jw) [inline utils/findMatchingRecord.js:13]: I like that you added these comments in here, it shows you're really cementing the understanding of how this all works together!

File: utils/airtable.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+import findMatchingRecord from './findMatchingRecord';
```
Comment (dustin-jw) [inline utils/airtable.js:1]: One suggestion for the file structure: we probably want the `utils` folder to be inside the `src` folder. Typically we want the project root to be mostly configuration stuff. `src` is where our main application code/business logic should live.

**Review by dustin-jw — CHANGES_REQUESTED** [review]
I mostly have nitpicks and organizational suggestions. This overall looks great and I'm psyched to see it coming together!

File: utils/airtable.js (hunk 14 above)
Comment (marissahuysentruyt) [reply utils/airtable.js]: Sure, that makes sense! 

Should be changed here: https://github.com/sparkbox/accessible-components/pull/34/commits/6f780decbc1db35ed3c1233dca50a4a9302016d1

File: utils/airtable.test.jsx (hunk 12 above)
Comment (marissahuysentruyt) [reply utils/airtable.test.jsx]: Easy fix! 
Does this look better? https://github.com/sparkbox/accessible-components/pull/34/commits/2c75dcfda2111c6bb615701ac4084aaff72919e8

File: src/pages/[id].page.jsx (hunk 9 above)
Comment (marissahuysentruyt) [reply src/pages/[id].page.jsx]: The `[id].page.test` was deleted, and then I removed that `throw error` comment, this `data-testid` that wasn't used. 
https://github.com/sparkbox/accessible-components/pull/34/commits/d6ebff5b80570b46f2d94f3ef5c124e08f070993

File: src/pages/[id].page.jsx (hunk 7 above)
Comment (marissahuysentruyt) [reply src/pages/[id].page.jsx]: https://github.com/sparkbox/accessible-components/pull/34/commits/0a92e87571835e5c17d53cc5a0832efd04b5d3c1

Better!

File: package.json (hunk 6 above)
Comment (marissahuysentruyt) [reply package.json]: https://github.com/sparkbox/accessible-components/pull/34/commits/d2c63ce28db633a9142299a38b395fe45882acee
delete

File: src/styles/components/_component-card.scss (hunk 10 above)
Comment (marissahuysentruyt) [reply src/styles/components/_component-card.scss]: Honestly, I think these changes should have already been in here, but with dropping those fix/style commits with the changes I made, I broke the stylesheet that was originally here. I can remove this commit if you want (I just gotta make sure I merged the branches in the right order to get this code back!)

**Review by bsmrdel101 — DISMISSED** [review]

**Review by changangus — DISMISSED** [review]
This PR looks great! Nice job Marissa!

**Review by dustin-jw — DISMISSED** [review]
Approved in triplicate, merge away!

**Review by dustin-jw — APPROVED** [review]
